Run all scripted tests using RunFromSourceMain

eatkins opened the pull request:
This PR does two big things:
1. Removes the launcher based scripted implementation and only used `RunFromSourceMain`. This has the benefit of not requiring recompilation of the compiler bridges every time. It also simplifies making new tests because it is no longer necessary to consider whether it is launcher or run from source based.
2. Use `publishLocalBin` instead of `publishLocal` so that we can avoid the expensive `doc` operation in `publishLocal`.

eatkins committed:
Fix RunFromSourceMain sbt.Package$ bug

The main reason for having both the RunFromSourceMain and LauncherBased
scripted tests was that RunFromSourceMain would fail for any test that
ended up accessing the sbt.Package$ object. This commit fixes this bug
by reworking the classloader generated by RunFromSourceMain to invoke
sbt, switching from the classpath to jar classpath (by setting exportJars =
true) and entering sbt by calling `new xMain().run` rather than
`xMain.run`.

The reason for switching to the jar classpath is that the jvm seems to
have issues when there are two classes provided in different directories
that have the same case insensitive name, e.g. `sbt.package$` and
`sbt.Package$`. If those classes are instead provided in different
jars, the jvm seems to be able to handle it.

Exporting the jars is not enough though, I had to rework the
ClassLoader created in the launch method to have a layout that was
recognized by xMainConfiguration. I reimplemented the AppConfiguration
in java so that it could bootstrap itself in a single jar classloader
(the only needed jar is the Scripted.

If we export the jars in the build, then the NoClassDefErrors for
`sbt.Package$` go away during scripted tests using RunSourceFromMain.
This might make running tests in subprojects slightly slower but I think
its a worthy tradeoff.

eatkins committed:
Remove launcher based scriped runner

After the changes to `RunFromSourceMain` and the addition of
`Compile / exportJars` to the build, all of the tests pass with
RunFromSourceMain so I decided to remove the launcher based ones. The
RunFromSourceMain based tests are faster because they don't recompile
the various compiler bridges every time.

eatkins committed:
Add test sources to scriptedUnpublished/watchTriggers

eatkins committed:
Explicitly set scripted and server test classpath

This commit makes it so that the scalaVersion, sbtVersion and classpath
are always passed in as parameters to any method that creates an sbt
server -- either for scripted or for the sbt server tests. By making
that change, I was able to change the implementation of scripted in the
sbt project to use publishLocalBin instead of publishLocal. This makes
the scripted tests start much faster (doc alone can easily take 30
second) with messing with the build to exclude slow tasks from
publishLocal.

As part of this change, I removed the test dependency on scriptedSbtRedux for
sbtProj and instead had scriptedSbtRedux depend on sbtProj. This allowed
me to remove some messy LocalProject logic in the resourceGenerators for
scriptedSbtReduxProj. I also had to remove a number of imports in the
scriptedSbtReduxProj because the definitions available in the sbt
package object became available.

I also removed the dependency on sbt-buildinfo and instead pass the
values from the build into test classes using scalatest properties. I
ran into a number of minor issues with the build info plugin, namely
that I couldn't get fullClasspathAsJars to reliably run as a BuildInfo
key. It also is somewhat more clear to me to just rely on the built in
scalatest functionality. The big drawback is that the scalatest
properties can only be strings, but that restriction isn't really a
problem here (strangely the TestData structure has a field configMap
which is effectively Map[String, Any] but Any is actually always String
given how the TestData is created as part of framework initialization.

Since scripted no longer publishes, scriptedUnpublished is now
effectively an alias for scripted.

To get publishLocalBin working, I had to copy private code from
IvyXml.scala into PublishBinPlugin. Once we publish a new version of
sbt, we can remove the copied code and invoke IvyXml.makeIvyXmlBefore
directly.

eatkins committed:
Allow RunSourceFromMain to use launcher

The repo overrides scripted test relies on using the launcher to modify
the default resolvers. To support this, I extended the scripted launcher
to use the bundled sbt launcher if it is passed in via the
`-Dsbt.launch.jar` system property.

dwijnand commented:
Oh, man, this is an exciting PR!

But I'm a little confused.  You mentioned the lack of having to publish (🎉), but then switching back to publish bin only...  Is publishing still required?  The ideal would be that compiling is only required (or `products`, perhaps).

dwijnand commented on build.sbt:
```diff
@@ -1202,18 +1158,20 @@ ThisBuild / scriptedPrescripted := { _ =>
-    publishLocalBinAll := { val _ = (publishLocalBin).all(ScopeFilter(inAnyProject)).value },
-    aggregate in bintrayRelease := false
+    publishLocalBinAll := Def.taskDyn {
+      (Compile / (if (insideCI.value) publishLocal else publishLocalBin)).all(scriptedProjects)
```
Not sure if this should vary in CI.  Also it makes the name `publishLocalBinAll` a bit misleading...

eed3si9n commented:
@dwijnand I initially had the same feeling, but then I realized that the metabuild still requires resolution of sbt core artifacts. I had a PR to exclude sbt core artifacts from the metabuild (https://github.com/sbt/sbt/pull/4589), but I had to add it back for Coursier (https://github.com/sbt/sbt/pull/4614/commits/f999f6a62ea89fc8292e4be20564652b3757a04e).

eatkins commented:
I think that @eed3si9n is right that some publishing is necessary. I don't think it's necessary to do publishLocal in ci though. I think the reason I added that was that I forgot that I pulled https://github.com/sbt/sbt/pull/5364 out of this PR so publishLocalBin didn't work in ci. `RunSourceFromMain` works fine with the artifacts from `packageBin` (which was my first approach) rather than `publishLocalBin` but sbt is unable to resolve itself without publishing. On my local machine, I generally had published ivy files for the sbt sub projects so that things would work with `packageBin` but tests would fail in ci.

eatkins commented:
If I do end up rewriting the launcher, one thing I will probably do is skip ivy resolution if the project sbt version matches the launcher version unless there is some reason I am unaware of for this behavior. I think this would generally make `packageBin` work in scripted unless we wanted to use scripted to test if a launcher can resolve a different launcher version. That kind of test probably shouldn't be a scripted test though.

eatkins committed:
Add extra classpath to the metabuild

When sbt was entered through xMain.run and the classloaders do not have
the expected format, sbt recreates the classloaders for itself.
Unfortunately the extra classpath was not added to the classloader. This
caused project/extra to fail if it was entered from RunFromSourceMain
rather than with the launcher.

eatkins committed:
Set jna.nosys=true in RunSourceFromMain

Without this, I saw errors on windows.

eatkins committed:
Fix reboot full in RunFromSourceMain

The initial implementation of reboot full did not work correctly. After
this change, the project/extra scripted test, which exercises reboot
full works.

eed3si9n commented:
For the record, we might need to flip back to using launcher when Scala version changes. It was straight forward to do once I found the `launch: Boolean` parameter - https://github.com/sbt/sbt/pull/5460

dwijnand commented:
> we might

Why?

eed3si9n commented:
I kept running into problem like this when I bumped the Scala version (I did it too in `project/plugins.sbt`):

```
[info] About to run tests:
[info]  * plugins/twirl
[info] Running 1 / 1 (100.00%) scripted tests with RunFromSourceMain
[info] Running plugins/twirl
[error] java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
[error] 	at sbt.internal.scriptedtest.ScriptedLauncher.launch(ScriptedLauncher.java:130)
[error] 	at sbt.RunFromSourceMain$.launch(RunFromSourceMain.scala:104)
[error] 	at sbt.RunFromSourceMain$.runImpl(RunFromSourceMain.scala:81)
[error] 	at sbt.RunFromSourceMain$.run(RunFromSourceMain.scala:69)
[error] btRoot /at sbt.RunFromSourceMain$.main(RunFromSourceMain.scala:57)
[error] 	at sbt.RunFromSourceMain.main(RunFromSourceMain.scala)
[error] Caused by: java.lang.reflect.InvocationTargetException
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.internal.scriptedtest.ScriptedLauncher.launchImpl(ScriptedLauncher.java:191)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.internal.scriptedtest.ScriptedLauncher.launch(ScriptedLauncher.java:104)
[error] 	... 5 more
[error] Caused by: java.lang.NullPointerException
[error] 	at sun.net.util.URLUtil.urlNoFragString(URLUtil.java:50)
[error] 	at sun.misc.URLClassPath.getLoader(URLClassPath.java:522)
[error] 	at sun.misc.URLClassPath.getNextLoader(URLClassPath.java:494)
[error] 	at sun.misc.URLClassPath.getResource(URLClassPath.java:248)
[error] 	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
[error] 	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
[error] 	at java.security.AccessController.doPrivileged(Native Method)
[error] 	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
[error] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
[error] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:405)
[error] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:405)
[error] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
[error] 	at sbt.internal.XMainConfiguration.run(XMainConfiguration.scala:70)
[error] 	at sbt.xMain.run(Main.scala:39)
[error] 	... 15 more
[error] java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
```

It's potentially possible that there are some assumption around build's Scala version being the same as the freshly baked sbt's Scala version somewhere, and we just need to fix it but I just flipped to using launcher scripted test.

eatkins commented:
Rather than switching to launcher, it would be much better to figure out the underlying issue with the new scala version. What version are you using that is causing problems?

eed3si9n commented:
I agree that it would be better to fix the root cause if that's possible. I just wanted to test Scala 2.12.11-bin-e1e2bfe quickly this morning, so I was also grateful of the escape hatch - https://github.com/sbt/sbt/pull/5460

dwijnand commented:
https://github.com/sbt/sbt/pull/5461

eed3si9n commented on scripted-sbt-redux/src/main/scala/sbt/scriptedtest/ScriptedTests.scala:
```diff
@@ -482,102 +403,62 @@ object ScriptedTests extends ScriptedRunner {
-    //new FullLogger(Logger.xlog2Log(log)))
-  }
-
-  // This is called by sbt-scripted 0.13.x and 1.x (see https://github.com/sbt/sbt/issues/3245)
-  def run(
-      resourceBaseDirectory: File,
-      bufferLog: Boolean,
-      tests: Array[String],
-      bootProperties: File,
-      launchOpts: Array[String],
-  ): Unit = {
-    val logger = TestConsoleLogger()
-    val prescripted = ScriptedTests.emptyCallback
-    run(resourceBaseDirectory, bufferLog, tests, logger, bootProperties, launchOpts, prescripted)
-  }
```
Because of `sbt ^scripted`, we need this entry point - https://github.com/sbt/sbt/issues/3245

eed3si9n commented on scripted-sbt-redux/src/main/scala/sbt/scriptedtest/ScriptedTests.scala:
```diff
@@ -482,102 +403,62 @@ object ScriptedTests extends ScriptedRunner {
-    runInParallel(baseDir, bufferLog, tests, bootProps, launchOpts, prescripted, 1)
-  }
-
   // This is used by sbt-scripted sbt 1.x
```
We need the following version for https://github.com/sbt/sbt/blob/2b557851ca168cd6274adafb7db5d9ed32efb5ee/main/src/main/scala/sbt/ScriptedPlugin.scala#L105-L120

```scala
  // This is used by sbt-scripted sbt 1.x
  def runInParallel(
      baseDir: File,
      bufferLog: Boolean,
      tests: Array[String],
      bootProps: File,
      launchOpts: Array[String],
      prescripted: java.util.List[File],
      instances: Int
  ): Unit = ...
```

https://github.com/sbt/sbt/issues/5592

eed3si9n replied:
I'll send a PR to recover these entry points.
